### DISCUSSION TITLE
How to achieve a "generic plural"?

### DISCUSSION THREAD

**Discussion by MichalBryxi** [ember-intl/ember-intl#1605] [Q&A] [answered]
## Problem definition

Thinking about cases where for plural forms it's not possible to say what the value for `count` is. Imagine a table where a column is expected to have a heading `comments` and the cells just show number telling you how many comments are there for the individual rows.

I would say that it makes sense to have a translation record looking like:

```
{count, plural, =1 {comment} other {comments}}
```

Now the question is how to request "a plural" in the code?

```hbs
<th>
  {{t "comment" count=???}}
</th>
```

What would be the value for `count`? `count=2` works in English, but for example for Czech it would need to be `count=5`. And possibly many other numbers for other languages. From the [ICU spec](https://support.crowdin.com/icu-message-syntax/#plural) it _seems_ like `other` is used in those cases?

## Question

So I guess the questions I have are:

- From language perspective, is there a rule that would say "when you don't know the count, use `two` / `few` / `many` / `other`?
- Whatever the correct form to use for generic plural is, is there a way to pass that down to the `t` helper? Something like: `{{t "comments" count="other"}}`

## Other option

Also I could potentially just have a plural form entry in the translation file and request that directly:

```hbs
{{t "comments"}}
```

It just feels off that I have to duplicate these every time and that there would no rule for such case?

**Comment by longlho** [comment]
I'm not sure I understand your question. First of all you shouldn't force `=1` but use `one` instead.
In terms of `{count, plural, =1 {comment} other {comments}}` what do you mean by request "a plural"? What's your expected input & output?

**Reply by gnclmorais** [reply]
@longlho, I think he means just the plural form of a word — and I know this might sound too English centric or something, so let me explain.

Imagine the following sentence: “The following table has: emails, names, and phone numbers”. Those three properties are in their (English) plural form, but possess no numbers to choose a plural from. I think that’s what Michal is looking for with a “generic plural”, a plural form that languages could use when talking about a numberless plural.

**Reply by longlho** [reply]
If you have a fixed sentence without any placeholder, the translator should be able to translate as is. In the example you mentioned `The following table has: emails, names, and phone numbers` since it's a constant it's translatable. The generic plural doesn't really exist since it depends on the default plural AND gender form of words in a language (e.g some lang by default assume gender/plurality of things to be male/plural while other female/singular depending on what that thing is).

**Reply by gnclmorais** [reply]
In my case, `emails, names, and phone numbers` are dynamic, i.e., they are “summaries” of fields that might (or not) be present on the data that we show after that sentence; so they are translated in separate.

**Reply by longlho** [reply]
So can you just declare `emails`, `names`, `phone numbers` as is with that context and translators take care of that? I don't think it relies on plural form per se.

**Reply by MichalBryxi** [reply]
To put the above example in code, I guess as a simplified code it can be something like:

```hbs

{{t "The following table has:}}

{{#each @types as |type|}}
  {{t type}}
{{/each}}
```

> So can you just declare emails, names, phone numbers as is 

I'm not sure what would that mean, but I read this as: Put `emails`, `names` and `phone numbers` directly as keys in the translation files? In other words the "other option" from my original question.

**Reply by longlho** [reply]
yeah, and join them w/ `ListFormat`

**Reply by MichalBryxi** [reply]
Oh ok. That sounds like a reasonable solution. Thanks for the discussion/explanation! 🌟  🙇🏾 

To put it into code for someone coming by:

```js
get typesString() {
  const typesTranslated = this.args.types.map((type) => this.intl.t(type));
  return new Intl.ListFormat('en', { style: 'long', type: 'conjunction' }).format(typesTranslated);
}
```

```hbs
{{t "The following table has: {types}" types=this.typesString}}
```

**Comment by MichalBryxi** [comment, marked as answer]
// Extracted from a discussion above:

> The generic plural doesn't really exist since it depends on the default plural AND gender form of words in a language (e.g some lang by default assume gender/plurality of things to be male/plural while other female/singular depending on what that thing is).

> You just declare `emails`, `names`, `phone numbers` as is with that context and translators take care of that. And join them w/ [ListFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat).

```js
// this.args.types = ['emails', 'names', 'phone numbers];
get typesString() {
  const typesTranslated = this.args.types.map((type) => this.intl.t(type));
  return new Intl.ListFormat('en', { style: 'long', type: 'conjunction' }).format(typesTranslated);
}
```

```hbs
{{t "The following table has: {types}" types=this.typesString}}
```
